### ISSUE TITLE
BertTokenizerFast.from_pretrained max_len does't work

### ISSUE THREAD

**Issue by moseshu** [huggingface/tokenizers#866]
config = BertConfig(
        vocab_size=10000,
        hidden_size=256,
        num_hidden_layers=6,
        num_attention_heads=4,
        intermediate_size=3072,
        hidden_act="gelu",
        hidden_dropout_prob=0.1,
        attention_probs_dropout_prob=0.1,
        max_position_embeddings=512,
        type_vocab_size=2,
        pad_token_id=0,
        position_embedding_type="absolute",
        truncation=True
    )

    tokenizer = BertTokenizerFast.from_pretrained('./tokenizer', max_len=512,
                                                  do_lower_case=False)
    tokenizer.add_special_tokens({'mask_token': '[MSK]'})

    # print(len(tokenizer))
    model = BertForPreTraining(config)
 train_dataset = TextDatasetForNextSentencePrediction(
        tokenizer=tokenizer,
        file_path='./movies.txt',
        block_size=512,
        overwrite_cache=False,
        short_seq_probability=0.1,
        nsp_probability=0.5,
    )
the above is my code ,the class BertTokenizerFast param max_len does't work. 
when I train my custom data ,the seq lenth is 893. 
they can't add in modeling_bert ("embeddings += position_embeddings").
 how to set he max_seq_length to limit the sentence length?

**Comment by Narsil** [comment]
Hi @moseshu ,

You BertConfig specifies `max_position_embeddings=512` are you sure about `893` ?
Usually the data just gets truncated to `512`, but you can definitely try to push it to `1024` to fit all your data (If you're doing finetuning this won't work since position embeddings are learned with `max_len=512` on original Bert) .

Then tokenizer and models are entirely decorelated entities, meaning they don't share anything in common. `max_len` is an attribute of the `model`. The equivalent is called `model_max_length` in `transformers`. If you really want to override you can do something like

```python
from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained('bert-base-uncased')
tokenizer.model_max_length = 1024
```
that should work. Again be careful about the interactions with the model.

**Comment by moseshu** [comment]
> 

thank you! I set the max_positional_embedding=1024 .the error is solved.
I try to save my custom tokenizer for AutoTokenizer or BertTokenizer use.
how to use BertWordPieceTokenizer save_model with config.json file ?
I use tokenizer = BertTokenizerFast.from_pretrained('./tokenizer') to reload the vocab.txt and then save_pretrained .
but there is no model_type in config.json file.

my code is 
`files = "./movie.txt"
    vocab_size = 10000
    min_frequency = 2
    limit_alphabet = 10000
    special_tokens = ["[PAD]", "[UNK]", "[CLS]", "[SEP]", "[MASK]"]  # for Bert and Albert

    # Initialize a tokenizer
    tokenizer = BertWordPieceTokenizer(clean_text=True, handle_chinese_chars=True, strip_accents=True, lowercase=True)

    # Customize training
    tokenizer.train(
        files,
        vocab_size=vocab_size,
        min_frequency=min_frequency,
        show_progress=True,
        special_tokens=special_tokens,
        limit_alphabet=limit_alphabet,
        wordpieces_prefix="##"
    )
    tokenizer.save_model("./tokenizer")


    `

**Comment by Narsil** [comment]
Yes, it seems it isn't saved properly when simply overriding like this and `.save_pretrained`.

Two possible fixes:

```python
tokenizer1 = BertTokenizerFast.from_pretrained('./tokenizer')
tokenizer1.model_max_length = 1024
tokenizer1.init_kwargs["model_max_length"] = 1024 # This is necessary because `save_pretrained` uses those values at init times, NOT the ones overridden afterwards)
tokenizer1.save_pretrained("./tokenizer") #tokenizer_config.json , special_tokens_map.json ,tokenizer.json

tokenizer2 = AutoTokenizer.from_pretrained("./tokenizer")
assert tokenizer2.model_max_length == 1024
``` 

Or 

Simple edit `tokenizer_config.json` and add `model_max_length: 1024` in it.
